Social Links: Add ability to change social icon sizes

mkaz opened the pull request:
## Description

This adds a size button control to the block toolbar, that lets you pick between 4 different sizes of icons (Small, Normal, Large, Huge). Those sizes affect the default style, the pill shape style, the logos only style, and even those variations of the new setup state as well.

Fixes #17276

## How has this been tested?

- (Backwards Compatibility) Create a post with social icons and confirm after applying the PR that they still can be edited and changed.
- Add social icons, adjust their sizes using the sidebar
- Confirm the new sizes look right in the editor and published.

## Screenshots

![social-link-size](https://user-images.githubusercontent.com/45363/95382604-bafe1200-089e-11eb-81c3-8a95621b1e75.gif)


## Types of changes

- Adds new attribute for size for Social Links
- Adds CustomSelect control to adjust sizes
- Adds custom CSS property for different sizes

mkaz committed:
Social Links: Add variables to change icon sizes

This adds CSS custom properties to alter the circle/logo sizes. I'm not
quite sure how to integrate this in to make it adjustable, maybe as a
theme.json optoin?

First, we may want to create as fixed sizes (sm, med, large) to make
sure things look right. There are two values to adjust, the outer circle
and the inner logo. So it is possible to make the icons all out of sorts
if the sizes don't match properly.

jasmussen commented:
This is what I see:

![socl](https://user-images.githubusercontent.com/1204802/95438246-50101200-0956-11eb-84cb-bd0ffd46c623.gif)

Nice work! This will be a nice feature.

As you can see from my inspecting, it needs a little visual love that I can help with — more adaptive centering, to be specific, so as to work with arbitrary sizes.

One larger question, though, and I'll defer to folks like @nosolosw @kjellr @iamtakashi @alaczek and others: should this be a global CSS variable for all social links icons? Or should it be a per-block size setting? I've no real strong opinion but would love insights.

nosolosw commented:
:wave: Took a quick look at this and it seems that you may be able to go with preset sizes (like small/medium/large) for the icons at the moment, is that correct? If so, I'd suggest making this work with classes instead, as the font size control does for presets.

If you also wanted custom values provided by the user, that's a bit more entangled and there's no clear solution yet for this kind of setup (where we want to set a property with a custom value in the container that affects all children). The issue is kind of similar to this other PR https://github.com/WordPress/gutenberg/pull/25372 The closest we've gone to address that use case is with the link control. The way it works there is by adding to the container a CSS variable with the custom value as an inline style. However, this approach is problematic with `kses` as it filters out any CSS variable within the style property of an element so it won't work for users without `unfiltered_html` permissions (author role, for example). See https://github.com/WordPress/gutenberg/pull/25411/ and https://github.com/Automattic/wp-calypso/issues/45456. The path forward seems to be either teach `kses` to allow any CSS variable within inline styles or look for an alternative approach.

kjellr commented:
> One larger question, though, and I'll defer to folks like @nosolosw @kjellr @iamtakashi @alaczek and others: should this be a global CSS variable for all social links icons?

I think a global size for all social links icons is adequate. As @nosolosw mentioned, I think a small/medium/large dropdown sort of setup (along with perhaps a custom value field) would be great.

jasmussen commented:
A set of presets is a good idea as it allows us to pick presets that render the SVG icons well. Here are some thoughts:

Huge (48px/2x):

<img width="222" alt="huge" src="https://user-images.githubusercontent.com/1204802/95551410-06cdca00-0a0b-11eb-8701-3f444e9f33c8.png">


_Probably too huge, probably won't be used that often. But doesn't hurt to have it. Maybe you want to make a statement._


Large (36px/1.5x):

<img width="206" alt="large" src="https://user-images.githubusercontent.com/1204802/95551536-4399c100-0a0b-11eb-96e6-6b473f2d77f6.png">


Big (32px/1.33_x) — best if we can do without this one as it's not a good scaling integer:

<img width="213" alt="32 maybe" src="https://user-images.githubusercontent.com/1204802/95551677-89568980-0a0b-11eb-822a-90d83644c4b2.png">


Default (24px/1x):

<img width="216" alt="default" src="https://user-images.githubusercontent.com/1204802/95551700-93788800-0a0b-11eb-929b-a72baa9a3ade.png">

Small (18px/.75x):

<img width="214" alt="small 18" src="https://user-images.githubusercontent.com/1204802/95551804-c458bd00-0a0b-11eb-9d14-0b347a385861.png">


Tiny (12px/.5x):

<img width="207" alt="tiny 12" src="https://user-images.githubusercontent.com/1204802/95551741-aa1edf00-0a0b-11eb-8886-438d815c567c.png">

__Perhaps too small, but it's a decent scaling integer_

What do you think?

kjellr commented:
Those sizes look good! I agree we can probably lose the "Big" one. The "Tiny" one is likely too small to recommend to users, so I think it's fine to remove it from the presets.

mkaz committed:
Switch social links to fixed sizes

Uses small, normal, large, and huge sizes matching the same presets as
the font sizes.

Uses the CustomSelectControl to pick which size

Introduces class names `has-SIZE-icon-size` that applies to social link

kjellr committed:
Ensure that icons aren't cut off by their button container in the editor.

jasmussen commented:
This is what I see:

![size](https://user-images.githubusercontent.com/1204802/96119303-135f9000-0eed-11eb-8f59-de34f6d4847d.gif)

Nice work, this will make a bunch of people happy.

However a couple of things — this control doesn't really fit in the sidebar, in part because it's a primary action for the block. It should just be a popover that opens from the block toolbar, it can just be a text button like this:

<img width="65" alt="Screenshot 2020-10-15 at 13 49 06" src="https://user-images.githubusercontent.com/1204802/96119426-41dd6b00-0eed-11eb-8ca7-d4bf00ad423f.png">

And the font size indication isn't really helping here, it seems fine to just do a menu like this:

<img width="333" alt="Screenshot 2020-10-15 at 13 51 34" src="https://user-images.githubusercontent.com/1204802/96119627-8f59d800-0eed-11eb-9e66-f9fb41b0a61f.png">

jasmussen commented on packages/block-library/src/social-link/editor.scss:
```diff
@@ -1,6 +1,7 @@
 .wp-block-social-links .wp-social-link button {
 	color: currentColor;
 	padding: 6px;
+	height: var(--wp-social-link-size);
```
Is it appropriate to surfaces this as a CSS variable even if it's not part of global styles? @nosolosw 

I think I'd prefer we avoid CSS variables entirely unless it's directly tied to global styles. An alternative is to have CSS classes, `is-small`, `is-large`, `is-huge` instead, and then apply the sizing based on that.

kjellr replied:
My expectation was that at some point themes might be able to modify this via global sizes — similar to how they can provide font size presets. 

That may be overkill from a design tools perspective, but I figured it can't hurt to allow that.

jasmussen replied:
Theme-abillity through global styles is definitely a nice feature, and if that requires these to be CSS variables (and if this is the correct way to do that), then I'm all thumbs up from here.

nosolosw replied:
I'd probably go with raw values instead of CSS variables in this PR. There's nothing wrong with using CSS vars as they are here (it's actually a nice use case). My only concern is that people start using them to tweak social icons so they become public API and we can't modify how this works down the road when/if this is integrated in Global Styles proper.

For some context: so far, the styles that can be declared via Global Styles are a straightforward 1:1 mapping to some CSS property (eg: fontSize, backgroundColor, etc). The needs here are a bit different in that four things need to be coordinated at once (width/height of the link, width/height of the logo), which makes that behaviour more particular to this specific block and less as a general utility (unlike fontSize or backgroundColor). This is not to say it can't be done, but that I see a few challenges that merit its own separate discussion.

mkaz replied:
Yes, the idea using CSS variables was that a theme would be able to provide a custom value or preset.

jasmussen commented:
Thanks so much for this PR, I know my feedback means a little extra work. But I think it will be worth it, and that doing it the right way will leave it without additional work necessary.

mkaz committed:
Switch size control to toolbar

mkaz commented:
I've moved the controls to the toolbar, test it out and let me know what you think. 👍

jasmussen commented:
Thank you, this is what I see:

![size](https://user-images.githubusercontent.com/1204802/96226759-6724b480-0f93-11eb-9c69-4a5c44e79dc1.gif)

This is definitely the right interaction, but the button doesn't look quite as fitting as I had hoped. I think we should still try it for now, but I may go back and change it to an icon separately, if I can find a good size icon.

There needs to be a checkmark on the right side of the selected size (normal being selected by default). There should be code for that in the more options dropdown:

<img width="310" alt="Screenshot 2020-10-16 at 09 40 15" src="https://user-images.githubusercontent.com/1204802/96226932-a0f5bb00-0f93-11eb-9fc1-ff1d9473d5ae.png">

jasmussen commented:
This is unlikely to make it to 5.6, correct? Asking because looking at the code, I'd like to refactor it a bit, both to address Andrés feedback about not using CSS variables, but also to make the padding scale more naturally to the size. Ideally there's only one uniform padding that automatically pads more for larger icon sizes and less for smaller ones (like a percentage or something that scales with a value). But in order for that to work smoothly with both circles and pill shapes, I'll need to be able to focus on it for a bit, and I won't have time to do that until Monday.

mkaz committed:
Add check icon for selected menuitem

mkaz committed:
Add isAlternate prop for Dropdown

mkaz commented:
@jasmussen I updated with the check icon and setting the `isAlternate` prop that gives the proper type of popup. Both of those changes should address what you had above.

I also rebased against the placeholder social links, but in the rebase I broke things and now have chubby icons. I figured you wanted to rework some of the CSS, and you'll end up doing it much better than me, I'll let you work your magic to get them right. Needed for both in editor and published view. Thanks!

paaljoachim commented:
My test shows this:

![Social-icons-size](https://user-images.githubusercontent.com/5323259/96366483-202bf000-1148-11eb-9b04-c34fc6a653f1.gif)

Btw
Should there also be a setting in the sidebar with a size drop down? Having it in the toolbar and the sidebar.

jasmussen commented:
I'm going to take a stab at bringing this home today!

Thanks for taking a look, Paal.

> Should there also be a setting in the sidebar with a size drop down? Having it in the toolbar and the sidebar.

I'd suggest no: redundancy is okay in places, but it's hard to get right. It always begs the question: if there are two ways to accomplish the same, which one is the right way? And in absence of the redundancy, the principle to evaluate against is whether the action is primary (should be in toolbar) or secondary (should be in sidebar), and I'd suggest this is definitely primary.

jasmussen committed:
Initial refactor work.

jasmussen committed:
Refactor to use ems.

jasmussen commented:
As I worked on this, and looking for a clear and scalable (for themes also) heuristic for what the padding is, I ended up with a basic system. 6x6 units. An icon is 4x4 units with 1unit of padding all around. So if an icon is 24px, then it has a 6px padding (24/6). 

That works nicely with ems —`font-size: 24px; width: 1em; height: 1em; padding: 0.25em;` — however it did surface one gotcha.

18/4 = 4.5. And half-pixels don't render well, which means if we went with 18px for the "small" icon size, icons wouldn't be centered. They'd have 5px top and left padding, and 4px bottom and right padding. 

So I changed the small size from 18 to 16. It looks alright:

<img width="346" alt="Screenshot 2020-10-19 at 10 21 37" src="https://user-images.githubusercontent.com/1204802/96420537-ea454500-11f5-11eb-906c-d9f2e6c6a0cc.png">

jasmussen commented:
@mkaz we need for the checkmark to be present on the "normal" size, even if you haven't explicitly set the property. I.e. insert a social links block, do nothing, check the size, and you see this:

<img width="540" alt="need a default" src="https://user-images.githubusercontent.com/1204802/96420627-034df600-11f6-11eb-92cc-4ba9225ff769.png">

That is — if you have no property explicitly set, the checkmark should also show on "normal".

jasmussen commented:
Just testing this EM approach with a few themes, and so far it seems alright:

2020 back and front:

<img width="476" alt="Screenshot 2020-10-19 at 10 30 44" src="https://user-images.githubusercontent.com/1204802/96420888-56c04400-11f6-11eb-853d-b5ecfce847a9.png">

<img width="452" alt="Screenshot 2020-10-19 at 10 30 51" src="https://user-images.githubusercontent.com/1204802/96420890-57f17100-11f6-11eb-9f8d-6becacd82014.png">

2019:

<img width="518" alt="Screenshot 2020-10-19 at 10 31 06" src="https://user-images.githubusercontent.com/1204802/96420898-5b84f800-11f6-11eb-8aa5-ae966caad46c.png">

<img width="401" alt="Screenshot 2020-10-19 at 10 31 16" src="https://user-images.githubusercontent.com/1204802/96420901-5cb62500-11f6-11eb-82b2-0b3a76f2e7e9.png">

That margin in 2019 in the editor issue, I believe, is present also in the main branch.

jasmussen committed:
Small tweaks.

jasmussen committed:
Fix variations.

jasmussen commented:
Got variations working. Frontend:

<img width="575" alt="front" src="https://user-images.githubusercontent.com/1204802/96423456-a6543f00-11f9-11eb-9a1d-af01df69370c.png">

Editor:

<img width="787" alt="editor" src="https://user-images.githubusercontent.com/1204802/96423464-a8b69900-11f9-11eb-8828-6471388d5176.png">

Before:

<img width="431" alt="Screenshot 2020-10-19 at 10 54 46" src="https://user-images.githubusercontent.com/1204802/96423475-aa805c80-11f9-11eb-8450-9c575dc9e0bb.png">

jasmussen committed:
Get the setup state working.

jasmussen commented:
Got the setup new state working, both with variations and size:

<img width="995" alt="Screenshot 2020-10-19 at 11 47 56" src="https://user-images.githubusercontent.com/1204802/96429273-0dc1bd00-1201-11eb-8e61-ec18c54ee002.png">

I think this one is actually ready to review.

jasmussen commented:
Forgot to mention, we need https://github.com/WordPress/gutenberg/pull/25921#issuecomment-711840866 fixed as well before this one can ship. Let me take a stab now.

jasmussen committed:
Don't size up placeholder state on hover.

draganescu commented on packages/block-library/src/social-link/editor.scss:
```diff
@@ -1,9 +1,20 @@
-	padding-left: 16px;
-	padding-right: 16px;
+	padding-left: 0.6666em;
+	padding-right: 0.6666em;
```
Now, this is a very specific number 😄

jasmussen replied:
Heh, yes. The math checks out for the ems.

When the font size is 24px, 0.6666666666666 (infinite 6es but there wasn't space to write them all out) comes to 16px, which is the padding the pill shape should have left and right.

kirilzh replied:
I think in this case you can do `calc((2/3) * 1em)`

jasmussen replied:
That's not bad! Let me try that, thank you.

kirilzh commented on packages/block-library/src/social-links/style.scss:
```diff
@@ -16,19 +16,56 @@
+
+	// Huge.
+	&.has-huge-icon-size {
+		font-size: 48px;
```
for the font-sizes you could do with size variables, since they are all multiples of 8

jasmussen replied:
I tend to avoid variables unless they are to be reused in a number of places, and especially in the style.scss files, as they are just meant for the frontend.

mkaz committed:
Add checkmark to normal size even if not set

The default size is normal, even if the iconsize is not set,
so this sets the checkmark in the dropdown to reflect.

jasmussen commented:
Oh nice, thank you @mkaz! Let me address the other feedback!

jasmussen committed:
Use math for nicer padding. Props @kirilzh

jasmussen commented:
Nice. The checkmark correctly shows up for normal as default, even if it isn't explicitly set:

<img width="591" alt="Screenshot 2020-10-19 at 15 31 17" src="https://user-images.githubusercontent.com/1204802/96458041-c0edde80-1220-11eb-9e9b-74427c8b3f2a.png">

And with the nice `calc` trick for the padding, the padding computes to the correct padding, as opposed to relying on rounding:

<img width="297" alt="Screenshot 2020-10-19 at 15 32 49" src="https://user-images.githubusercontent.com/1204802/96458095-d4994500-1220-11eb-8013-30e92b16b573.png">

jasmussen committed:
Add specificity to margins to fix a theme issue.

jasmussen commented:
In https://github.com/WordPress/gutenberg/pull/25921/commits/d4b71b790fdb3df25dedf73927eb7dacd157b44e I pushed a fix that adds specificity to the margin classes for social icons. While this is not the prettiest fix, it is still overridable if a theme has a good reason to do so, and it fixes an issue that is present in 2019. Before:

<img width="835" alt="Screenshot 2020-10-19 at 19 00 50" src="https://user-images.githubusercontent.com/1204802/96488647-9a3ea080-123e-11eb-8e37-9d0f5440003b.png">

After:

<img width="617" alt="Screenshot 2020-10-19 at 19 06 03" src="https://user-images.githubusercontent.com/1204802/96488655-9c086400-123e-11eb-85c9-8add88375029.png">

mtias commented on packages/block-library/src/social-links/block.json:
```diff
@@ -6,6 +6,9 @@
 			"type": "boolean",
 			"default": false
+		},
+		"iconSize": {
```
Does this attribute need the icon prefix or could it be just "size"?

mkaz replied:
Switched by Joen, thanks Joen!

mtias commented on packages/block-library/src/social-links/edit.js:
```diff
@@ -1,43 +1,124 @@
 		<div className="wp-block-social-links__social-placeholder">
-			<div className="wp-block-social-link wp-social-link-facebook"></div>
-			<div className="wp-block-social-link wp-social-link-twitter"></div>
-			<div className="wp-block-social-link wp-social-link-instagram"></div>
```
why is the `block-` removed?

jasmussen replied:
The frontend markup has the following CSS classes applied:

`<li class="wp-social-link wp-social-link-wordpress wp-block-social-links has-large-icon-size is-style-pill-shape">`

In effort to simply get the frontend and editor closer, it made sense to me to keep the classes more closely to the frontend.

mtias replied:
Yes, I'm confused about when the block prefix is added and when excluded

jasmussen replied:
I cleaned up a reference to `wp-block-social-link` so all CSS now refers to only `wp-social-link`. I'm not quite sure why we have both those classes, I'll defer to @mkaz.

nosolosw replied:
FWIW, the `wp-block-social-links` is only in children (li elements) because of a bug we just fixed https://github.com/WordPress/gutenberg/pull/26344

mkaz replied:
Marks this as resolved, per Andres comment.

mtias commented on packages/block-library/src/social-links/edit.js:
```diff
@@ -1,43 +1,124 @@
 		__experimentalAppenderTagName: 'li',
 	} );
+
+	const openOnArrowDown = ( event ) => {
```
Mmm, why is this needed here? We should not have these kind of definitions in the blocks themselves but rely instead on the toolbar components handling it

mkaz replied:
The event is fired when pressing the down arrow on the button, this is not the normal behavior for a button, but is natural (expected) in this case since it is opening a Dropdown menu, it will open and continuing down arrow will select the items.

I can see the need for a new type of button here that includes a dropdown, this behavior is also used in the Heading block for selecting level.

Testing again, and now it doesn't work as expected. The down arrow does open the menu, but pressing again does not move the selector, it loses focus.

mtias replied:
We have many controls in the toolbar that operate like dropdowns, seems it should be just handled with the core components. cc @diegohaz if you can give it a quick look

diegohaz replied:
@mkaz The generic `Dropdown` component doesn't add the key down handler automatically. `DropdownMenu` does. Have you tried it? For reference, the block switcher button uses it: https://github.com/WordPress/gutenberg/blob/99d02c51487c992699eab87c1c7ace90284d60e0/packages/block-editor/src/components/block-switcher/index.js#L159-L178

In case it helps, there's also this simple example on the developer handbook: https://developer.wordpress.org/block-editor/components/toolbar-item/#render-prop

mkaz replied:
Thanks @diegohaz, I've got it working using `DropdownMenu` but some of the same issues persist with that component, also seen elsewhere, so I created a ticket to track. See #26401

jasmussen committed:
iconSize -> size

jasmussen committed:
Hover cleanup.

paaljoachim commented:
This is what I see:
![Screen Shot 2020-10-21 at 12 25 18](https://user-images.githubusercontent.com/5323259/96707704-83ac5c80-1398-11eb-8d08-be742172e0b5.png)

Size selected is Large. Style: Default. 


Frontend:
![Screen Shot 2020-10-21 at 12 26 12](https://user-images.githubusercontent.com/5323259/96707842-ae96b080-1398-11eb-9896-cf721375723a.png)

Default and Pill Shape both show the above CSS styling.

jasmussen commented:
_shudder_ — those circles are ovals! 🙈  — I'll take a look.

jasmussen commented:
@paaljoachim I can't reproduce. This is what I see in twenty twenty:

<img width="910" alt="Screenshot 2020-10-21 at 13 16 22" src="https://user-images.githubusercontent.com/1204802/96712647-bc9bff80-139f-11eb-9dce-f426479d0f5d.png">

Can you try and make sure your cache is thoroughly cleared? Or are there other steps to reproduce?

paaljoachim commented:
I updated npm in terminal by adding the command: `npm install`
Restarted terminal and Docker.
Then added the following:
`git fetch origin pull/25921/head`
`git checkout 5aef47571ce2bc6cfae8bb17d2fe721b758ecf15`
`npm run build`

Restarted and updated:
`wp-env start --update`

(Also run `git reset --hard`)

Opened http://localhost:8888/ in Chrome. Did a hard refresh of Chrome.

If the newest is supposed to show by default a checkmark next to the normal size, then I am for whatever reason still seeing an older commit. Which would probably explain the oval circles.

jasmussen commented:
Two ways to make sure you're seeing the latest:
```
git checkout master
git pull
git branch -D [branch name]
git checkout [branch name]
npm install
npm run dev
```

then clear cache in your browser.

paaljoachim commented:
The newest commit is working properly.

Backend:
![Screen Shot 2020-10-21 at 14 37 50](https://user-images.githubusercontent.com/5323259/96720437-0d652580-13ab-11eb-9b3c-0d35cb8b0a18.png)

Frontend:
![Screen Shot 2020-10-21 at 14 38 34](https://user-images.githubusercontent.com/5323259/96720491-21a92280-13ab-11eb-98cb-0042c744b062.png)

jasmussen commented:
Phew! Thanks!

jasmussen commented:
Should we try changing the "Size" label in the toolbar to show what's selected? Since it's essentially a dropdown just like Alignments, and alignments show the alignment chosen, we could do the same here. We'd just want to change "Normal" to "Medium".

paaljoachim commented:
Yeah. I was thinking the same thing, but became hesitant that the selected size word would be seen and not the word "Size".

Here is an example of what it would look like.
![Social-Icons-Block-Selected-size](https://user-images.githubusercontent.com/5323259/96886206-a6676f80-1483-11eb-92c1-4ae2f73985bd.jpg)

mkaz committed:
Try using DropdownMenu component

mkaz commented:
@jasmussen I updated using the DropdownMenu component, there are still a few issues related to that component created a tracking ticket here #26401 — the biggest change for this ticket is that the DropdownMenu component only allows a button and not text in the toolbar.

Joen, try it out and let me know what you think, I'm not sure if there is a good size icon to use, otherwise we can revert to the previous and address after the base components improve.

jasmussen committed:
Fix calc.

jasmussen commented:
> the biggest change for this ticket is that the DropdownMenu component only allows a button and not text in the toolbar.

<img width="597" alt="Screenshot 2020-10-23 at 09 27 29" src="https://user-images.githubusercontent.com/1204802/96968934-03116b80-1512-11eb-9dc9-b08023709424.png">

Thanks so much for the work, but I can't see this working as well as the label. To me it's about this being a primary action for the block, therefore deserving not just a bigger button, but a label.

Is there any way we can extend or refactor the DropdownMenu to allow a label? Otherwise, what does the Image "Replace" button do to enable a dropdown from the label?

mtias commented:
I think the image replace dropdown was a custom implementation and it has some keyboard issues still.

jasmussen commented:
Argh. I just don't thin there's a good enough visual concept for "Size" to work as an icon.

mkaz committed:
Try text for button instead of icon

A test to see what we think of modifying the DropdownMenu/Button to
accept a text property which when specified and icon=null than it will
show instead of the icon.

mkaz commented:
@joen @mtias I pushed up a change to test out a modification to `DropdownMenu` and `Button` to accept a `text` field and if `icon` is specified as null, the text will show instead of an icon.

This would be moved to its own PR but pushed up here as a way to test and try out.

jasmussen commented:
Nice work. This is what I see:

![social](https://user-images.githubusercontent.com/1204802/97017550-5b665e80-154e-11eb-9b19-ea6b8bbcd2e2.gif)

Conceptually, that's perfect for me, except for two things:

- Small, normal, and large appear to not work anymore
- The dropdown menu itself should have the `isAlternate` prop to ensure the dark borders.

mkaz committed:
Add popover props for isAlternate

jasmussen commented:
I rebased this to see if it will sate the e2e tests.

I'm seeing a thing in the branch at the moment, specifically that the sizes don't change when I choose a different one. It appears we have duplicate size classes now:

![switch](https://user-images.githubusercontent.com/1204802/97145223-259bc280-1766-11eb-8d04-554c3c1d801a.gif)

Note how the initial saved class stays, and an additional class is added.

Finally, it really would be nice to show the sizes as small, medium, large, huge, directly in the button label. With those two fixed, it will be an awesome PR.

jasmussen commented:
Nice, this is working well for me:

![social links](https://user-images.githubusercontent.com/1204802/97855327-e12a9c80-1cfa-11eb-9662-3d82c2620acd.gif)

One more change and I think this is good to go.

Labels should be Small, Medium, Large, Huge, and I think we should make them show up as the label in the toolbar. Is this workable?

mapk commented:
Normally we've had the popover selection reflected in the block's toolbar (ie. alignment, position of text in Cover block, inline text color, etc ). I'm guessing representing the size with an icon is too ambiguous and showing the text of the size gets too wordy here?

mkaz committed:
Display size name as label

Switches the button to display the name as the label for example "Large"
instead of just showing "Size"

Fix issue with { onClose } not being destructured

Pass toggleProps forward.

mkaz commented:
@jasmussen Updating with label in button. See gif or try on the PR.

@mapk does this address your comment too?

![social-link-size-button](https://user-images.githubusercontent.com/45363/97902335-4d54d280-1cf2-11eb-81c4-615a4ae71720.gif)

mapk commented:
>@mapk does this address your comment too?

@mkaz thanks for trying it out so quickly! 
Showing the size in the toolbar, without any relationship to the other sizes, doesn't seem all that helpful IMO now that I'm seeing it in the gif.
I've also noticed that our text labels in the toolbar often indicate an action. "Large" or the other labels are descriptions. "Size" works pretty well because I really can't think of a better alternative. :) 
Sorry I'm not more helpful here. As you had it is probably best. I'll think on this more though. 👍

jasmussen commented:
Ugh, it seems that the popover component centers itself according to the width of the toolbar control (which makes sense if you think about it), but that means that any time you change a size property and the toolbar updates, the popover shifts position.

In absence of being able to fix that so the popover stays still, let's go back to the toolbar saying "Size". Then we can revisit if/when it becomes pressing in the future. 

Let's keep the labels as small, medium, large, huge. "Normal" is in the eye of the beholder anyway!

mkaz commented:
> Let's keep the labels as small, medium, large, huge. "Normal" is in the eye of the beholder anyway!

I initially was going to go with Medium, but looking at the font control it uses the "Normal" label, so for consistency I matched it, fine with changing either way.

mkaz committed:
Switch back to 'Size' in toolbar

jasmussen reviewed (approved):
One issue remains. Transforms are on two rows in this PR:

<img width="332" alt="Screenshot 2020-11-06 at 16 05 23" src="https://user-images.githubusercontent.com/1204802/98381658-7951ba00-204a-11eb-82a3-85f908cabc7d.png">

However a rebase fixes that, and the issue is not present in the master branch:

<img width="548" alt="Screenshot 2020-11-06 at 16 09 06" src="https://user-images.githubusercontent.com/1204802/98381690-866ea900-204a-11eb-9bf8-5709067b1b38.png">

I suspect that an issue was created and fixed in the master branch, while this branch was ongoing.

But since rebasing will likely cause those beautiful green checks to fail a random amount of times, I'd say it's okay to merge this one if you first test whether you see the same issue as me, and that a rebase (even if not pushed) fixes it.
